mysql-test/suite/binlog/t/binlog_recover_async_rollback_trx.test

```diff
@@ -0,0 +1,149 @@
+--let $restart_parameters=
+--source include/start_mysqld.inc
+--let $wait_condition= SELECT count(*) = 0 FROM information_schema.innodb_trx
+--source include/wait_condition.inc
+
+# Check whether the DML trx has been rolled back asynchronously.
+--let $assert_file= $MYSQLTEST_VARDIR/log/mysqld.1.err
+--let $assert_text= Server reported: Rolled back recovered transaction
+--let $assert_select= Rolled back recovered transaction
+--let $assert_count= 1
+--source include/assert_grep.inc
+
+# The transaction was set to ACTIVE in previous recovery, thus it should not
+# find any prepared thransactions.
+--let $assert_file= $MYSQLTEST_VARDIR/log/mysqld.1.err
+--let $assert_text= Server NOT report: Found 1 prepared transaction
+--let $assert_select= Found 1 prepared transaction
+--let $assert_count= 0
+--source include/assert_grep.inc
+
+UPDATE t1 SET data = data + 1 WHERE id = 0;
+--echo # expected (0, 2), (1, 1)
+SELECT * FROM t1;
+
+--echo # 3. Test that innodb shutdown as expected if any error happens before
+--echo #    normal rollback task is started. In the situation, rollback task
+--echo #    should be started at preshutdown accordingly to rollback or
+--echo #    deregister all recovered active transactions.
+# Generate an large active transaction
+--let $i = 2
+BEGIN;
+while ($i < 8)
+{
+  eval INSERT INTO t1 SELECT id + power(2, $i), 1 FROM t1;
+  --inc $i
+}
+COMMIT;
+
+BEGIN;
+UPDATE t1 SET data = 10;
+
+# Make sure above update is persisted.
+SET GLOBAL innodb_log_checkpoint_now = 1;
+--source include/kill_mysqld.inc
+
+# tc-heuristic-recover triggers an error before innodb rollback task start
+# Rollback task will not be started at preshutdown of read only mode. Active
+# transactions are not expected to rollback.
+--error 1
+--exec $MYSQLD_LAST_CMD --tc-heuristic-recover=ROLLBACK --innodb-read-only --log-error=$MYSQLTEST_VARDIR/tmp/log.err
+
+# Rollback task will not be started at preshutdown if recovery mode is greater
+# to 2.  Active transactions are not expected to rollback.
+--error 1
+--exec $MYSQLD_LAST_CMD --tc-heuristic-recover=ROLLBACK --innodb-force-recovery=3 --log-error=$MYSQLTEST_VARDIR/tmp/log.err
+
+# Rollback task will be started at preshutdown of fast shutdown if force
+# recovery is 2. But the transaction is deregistered instead of rollback.
+--error 1
+--exec $MYSQLD_LAST_CMD --tc-heuristic-recover=ROLLBACK --innodb-fast-shutdown=1 --innodb-force-recovery=2 --log-error=$MYSQLTEST_VARDIR/tmp/log1.err
+
+# Rollback task will be started at preshutdown of fast shutdown. The
+# active transaction is rolled back.
+--error 1
+--exec $MYSQLD_LAST_CMD --tc-heuristic-recover=ROLLBACK --innodb-fast-shutdown=1 --log-error=$MYSQLTEST_VARDIR/tmp/log1.err
+
+--source include/start_mysqld.inc
+
+--let $assert_file= $MYSQLTEST_VARDIR/tmp/log.err
+--let $assert_text= Server NOT report: Start rollback task at preshutdown
+--let $assert_select= Start rollback task at preshutdown
+--let $assert_count= 0
+--source include/assert_grep.inc
```

Comment:
I would find the following more concise and understandable:
```sql
--let SEARCH_FILE=$MYSQLTEST_VARDIR/log/mysqld.1.err
--let SEARCH_PATTERN=InnoDB: Start rollback task at preshutdown
--source include/shutdown_mysqld.inc
```
However, even better than checking the presence of such a message which does not seem to add much value to a normal user, I think that it could be better to start the server in `innodb_read_only=1` mode and check that the data is still there. You can find some ideas in `mysql-test/suite/innodb/t/innodb_force_recovery_rollback.test` and `mysql-test/suite/innodb/t/read_only_recover_committed.test`.

Reply:
yes, removed the warning and the assertions.